In [2]:
import pandas as pd

In [3]:
know_2017 = pd.read_csv('../know_Data/train/KNOW_2017.csv')

In [4]:
know_2017

idx  aq1_1 aq1_2  aq2_1 aq2_2  aq3_1 aq3_2  aq4_1 aq4_2  aq5_1  ...  \
0        0      3     3      3     3      3     3      4     4      3  ...   
1        1      4     5      4     5      3     4      3     4      3  ...   
2        2      3     4      3     4      3     4      5     6      4  ...   
3        3      3     3      3     3      3     5      4     5      4  ...   
4        4      4     5      3     4      3     4      4     5      3  ...   
...    ...    ...   ...    ...   ...    ...   ...    ...   ...    ...  ...   
9481  9481      3     5      2     4      3     3      2     2      2  ...   
9482  9482      5     5      5     5      5     5      3     4      4  ...   
9483  9483      3     3      4     6      3     3      4     5      4  ...   
9484  9484      3     5      3     5      4     5      3     4      3  ...   
9485  9485      3     4      3     4      3     4      3     4      3  ...   

     bq37  bq38  bq38_1  bq39_1 bq39_2  bq40 bq41_1  bq41_2 bq41_3  knowcode  
0      52     2      실업       1      1     1   4000           2200    825101  
1      38     4    건축공학       1      1     1                  2400    140204  
2      50     4    건축공학       1      1     1   4000           2400    140204  
3      42     4    환경학과       1      1     1   7000           3500    140601  
4      51     4    건축공학       1      1     1   4000           2500    140204  
...   ...   ...     ...     ...    ...   ...    ...     ...    ...       ...  
9481   50     4   산업디자인       1      1     1   5200           1800    411301  
9482   37     4  우주항공공학       1      1     1   4000           3000    151105  
9483   32     2      기계       1      4     2   2700           1500    701101  
9484   40     4   문헌정보학       1      1     1   6800           2500     25402  
9485   48     3      전기       1      1     1   4000           3700     15201  

[9486 rows x 156 columns]

## 이상치 파악

In [5]:
know_2017.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9486 entries, 0 to 9485
Data columns (total 156 columns):
 #    Column    Dtype  
---   ------    -----  
 0    idx       int64  
 1    aq1_1     int64  
 2    aq1_2     object 
 3    aq2_1     int64  
 4    aq2_2     object 
 5    aq3_1     int64  
 6    aq3_2     object 
 7    aq4_1     int64  
 8    aq4_2     object 
 9    aq5_1     int64  
 10   aq5_2     object 
 11   aq6_1     int64  
 12   aq6_2     object 
 13   aq7_1     int64  
 14   aq7_2     object 
 15   aq8_1     int64  
 16   aq8_2     object 
 17   aq9_1     int64  
 18   aq9_2     object 
 19   aq10_1    int64  
 20   aq10_2    object 
 21   aq11_1    int64  
 22   aq11_2    object 
 23   aq12_1    int64  
 24   aq12_2    object 
 25   aq13_1    int64  
 26   aq13_2    object 
 27   aq14_1    int64  
 28   aq14_2    object 
 29   aq15_1    int64  
 30   aq15_2    object 
 31   aq16_1    int64  
 32   aq16_2    object 
 33   aq17_1    int64  
 34   aq17_2    object 
 35 

In [6]:
# know_2017.isnull().sum()[:50]

# 2017은 null로 존재하는 값이 없다

In [7]:
for i in know_2017.columns : 
    print(i, len(know_2017[i].value_counts()))

idx 9486
aq1_1 5
aq1_2 8
aq2_1 5
aq2_2 8
aq3_1 5
aq3_2 8
aq4_1 5
aq4_2 8
aq5_1 5
aq5_2 8
aq6_1 5
aq6_2 8
aq7_1 5
aq7_2 8
aq8_1 5
aq8_2 8
aq9_1 5
aq9_2 8
aq10_1 5
aq10_2 8
aq11_1 5
aq11_2 8
aq12_1 5
aq12_2 8
aq13_1 5
aq13_2 8
aq14_1 5
aq14_2 8
aq15_1 5
aq15_2 8
aq16_1 5
aq16_2 8
aq17_1 5
aq17_2 8
aq18_1 5
aq18_2 8
aq19_1 5
aq19_2 8
aq20_1 5
aq20_2 8
aq21_1 5
aq21_2 8
aq22_1 5
aq22_2 8
aq23_1 5
aq23_2 8
aq24_1 5
aq24_2 8
aq25_1 5
aq25_2 8
aq26_1 5
aq26_2 8
aq27_1 5
aq27_2 8
aq28_1 5
aq28_2 8
aq29_1 5
aq29_2 8
aq30_1 5
aq30_2 8
aq31_1 5
aq31_2 8
aq32_1 5
aq32_2 8
aq33_1 5
aq33_2 8
aq34_1 5
aq34_2 8
aq35_1 5
aq35_2 8
aq36_1 5
aq36_2 8
aq37_1 5
aq37_2 8
aq38_1 5
aq38_2 8
aq39_1 5
aq39_2 8
aq40_1 5
aq40_2 8
aq41_1 6
aq41_2 8
bq1 21
bq2 6
bq3 6
bq4 2
bq4_1a 1600
bq4_1b 581
bq4_1c 180
bq5 2
bq5_1 7
bq5_2 3848
bq6 7
bq7 7
bq8_1 5
bq8_2 5
bq8_3 5
bq9 5
bq10 5
bq11 5
bq12_1 5
bq12_2 6
bq12_3 6
bq12_4 6
bq12_5 5
bq13 5
bq14 5
bq15_1 5
bq15_2 5
bq15_3 5
bq16 5
bq17 5
bq18_1 5
bq18_2 5
bq18_3 5
bq18


1. 이상치 파악 

aq41_1 자원관리 6 -> 값 33으로 나와있음 : 3으로 수정 


2. 주관식

살릴 주관식중 없다, 없음 -> null or 0 처리 

자격증 순서/ b,c 처리방법 
bq4_1a 자격증 1600
bq4_1b 자격증 581
bq4_1c 자격증 180


버림
bq5_2 훈련사항 3848 
bq19_1 향후일자리변화 8753 
bq30 유사직업명1199
bq31 사용하는도구 6778
bq32 직전직업 1569
bq33 전직가능한직업 1419
bq34 신직업발굴 340


bq38_1 학과 1453


3. 숫자여야하는 값

bq23 직업세계변화 53
bq37 나이 63
bq41_1 근로소득 216
bq41_2 근로소득 96
bq41_3 근로소득 148


4. 체크해봐야할값 - 어떻게 처리해야할지 고민 
bq39_1 고용형태 2
bq39_2 고용형태 7
bq40 계약형태 3

## 이상치 처리

In [8]:
for col in know_2017:
        know_2017[col].replace(' ', 0, inplace = True)


In [9]:
# know_2017.loc[:,:"aq41_2"].columns

In [10]:
# for col in know_2017.loc[:,:"aq41_2"].columns :
#     know_2017[col]=know_2017[col].astype(int)

In [11]:
know_2017.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9486 entries, 0 to 9485
Data columns (total 156 columns):
 #    Column    Dtype  
---   ------    -----  
 0    idx       int64  
 1    aq1_1     int64  
 2    aq1_2     object 
 3    aq2_1     int64  
 4    aq2_2     object 
 5    aq3_1     int64  
 6    aq3_2     object 
 7    aq4_1     int64  
 8    aq4_2     object 
 9    aq5_1     int64  
 10   aq5_2     object 
 11   aq6_1     int64  
 12   aq6_2     object 
 13   aq7_1     int64  
 14   aq7_2     object 
 15   aq8_1     int64  
 16   aq8_2     object 
 17   aq9_1     int64  
 18   aq9_2     object 
 19   aq10_1    int64  
 20   aq10_2    object 
 21   aq11_1    int64  
 22   aq11_2    object 
 23   aq12_1    int64  
 24   aq12_2    object 
 25   aq13_1    int64  
 26   aq13_2    object 
 27   aq14_1    int64  
 28   aq14_2    object 
 29   aq15_1    int64  
 30   aq15_2    object 
 31   aq16_1    int64  
 32   aq16_2    object 
 33   aq17_1    int64  
 34   aq17_2    object 
 35 

In [12]:
# 주관식문항 

str_li =['bq4_1a','bq4_1b','bq4_1c','bq5_2', 'bq19_1','bq30','bq31','bq32','bq33','bq34','bq38_1']

In [13]:
for i in know_2017.columns : 
    if i not in str_li : 
        know_2017[i] = know_2017[i].astype(int)
    else : 
        know_2017[i] = know_2017[i].astype(str)

In [14]:
know_2017.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9486 entries, 0 to 9485
Data columns (total 156 columns):
 #    Column    Dtype 
---   ------    ----- 
 0    idx       int32 
 1    aq1_1     int32 
 2    aq1_2     int32 
 3    aq2_1     int32 
 4    aq2_2     int32 
 5    aq3_1     int32 
 6    aq3_2     int32 
 7    aq4_1     int32 
 8    aq4_2     int32 
 9    aq5_1     int32 
 10   aq5_2     int32 
 11   aq6_1     int32 
 12   aq6_2     int32 
 13   aq7_1     int32 
 14   aq7_2     int32 
 15   aq8_1     int32 
 16   aq8_2     int32 
 17   aq9_1     int32 
 18   aq9_2     int32 
 19   aq10_1    int32 
 20   aq10_2    int32 
 21   aq11_1    int32 
 22   aq11_2    int32 
 23   aq12_1    int32 
 24   aq12_2    int32 
 25   aq13_1    int32 
 26   aq13_2    int32 
 27   aq14_1    int32 
 28   aq14_2    int32 
 29   aq15_1    int32 
 30   aq15_2    int32 
 31   aq16_1    int32 
 32   aq16_2    int32 
 33   aq17_1    int32 
 34   aq17_2    int32 
 35   aq18_1    int32 
 36   aq18_2    in

In [15]:
# know_2017.iloc[3985].aq41_1 = 3

In [16]:
know_2017.iloc[3985, 81] = 3

In [17]:
know_2017.iloc[3985, 81]

3

In [18]:
know_2017[(know_2017.aq11_1.isin([2,3,4,5])) & (know_2017.aq11_2==0)]['aq11_2']

89      0
577     0
2422    0
4568    0
Name: aq11_2, dtype: int32

In [19]:
know_2017.iloc[[89,577,2422,4568], 22] = 4 

In [20]:
know_2017[(know_2017.aq14_1.isin([2,3,4,5])) & (know_2017.aq14_2==0)]['aq14_2']

3137    0
4564    0
6539    0
Name: aq14_2, dtype: int32

In [21]:
know_2017.iloc[[3137,4564,6539], 28]=4

In [22]:
know_2017[(know_2017.aq27_1.isin([2,3,4,5])) & (know_2017.aq27_2==0)]['aq27_2']

715     0
792     0
3960    0
4060    0
5388    0
6383    0
9449    0
Name: aq27_2, dtype: int32

In [23]:
know_2017.iloc[[715,792,3960,4060,5388,6383,9449], 54]=4

In [24]:
know_2017[(know_2017.aq32_1.isin([2,3,4,5])) & (know_2017.aq32_2==0)]['aq32_2']

1458    0
2680    0
3523    0
3656    0
5301    0
6368    0
6711    0
6864    0
7435    0
9316    0
9319    0
Name: aq32_2, dtype: int32

In [25]:
know_2017.iloc[[1458,2680,3523,3656,5301,6368,6711,6864,7435,9316,9319], 64]=4

In [26]:
know_2017[(know_2017.aq41_1.isin([2,3,4,5])) & (know_2017.aq41_2==0)]['aq41_2']

1798    0
6894    0
7118    0
8289    0
Name: aq41_2, dtype: int32

In [27]:
know_2017.iloc[[1798,6894,7118,8289], 82]=4

In [28]:
null_li = know_2017[(know_2017.bq39_1==1) & (know_2017.bq41_1==0)]['bq41_1'].index

In [29]:
know_2017.iloc[:,152]

0       4000
1          0
2       4000
3       7000
4       4000
        ... 
9481    5200
9482    4000
9483    2700
9484    6800
9485    4000
Name: bq41_1, Length: 9486, dtype: int32

In [30]:
for i in null_li:
     know_2017.iloc[i,152]=round(know_2017[(know_2017.bq1 ==know_2017.iloc[i,83])&(know_2017.bq41_1!=0)].bq41_1.median())

In [31]:
null_li = know_2017[(know_2017.bq39_1==1) & (know_2017.bq41_1==0)]['bq41_1'].index
null_li

Int64Index([], dtype='int64')

In [32]:
null_li2 =know_2017[(know_2017.bq39_1==1) & (know_2017.bq41_3==0)].index

In [33]:
for i in null_li2:
    know_2017.iloc[i,154]=round(know_2017[(know_2017.bq1 ==know_2017.iloc[i,83])&(know_2017.bq41_3!=0)].bq41_3.median())

In [34]:
null_li2 =know_2017[(know_2017.bq39_1==1) & (know_2017.bq41_3==0)].index

In [35]:
null_li2 

Int64Index([], dtype='int64')

In [36]:
null_li3 = know_2017[(know_2017.bq39_1==2) & (know_2017.bq41_2==0)].index
null_li3 

Int64Index([ 466,  664,  810,  876,  877,  907, 1030, 1195, 1533, 1575, 1695,
            2003, 2482, 2597, 2738, 3331, 3818, 4025, 4153, 4669, 4687, 4738,
            5131, 5361, 5403, 6362, 6483, 6516, 6591, 7446, 7455, 7627, 7973,
            8075, 8224, 9080, 9341],
           dtype='int64')

In [37]:
for i in null_li3:
    know_2017.iloc[i,153]=round(know_2017[(know_2017.bq1 ==know_2017.iloc[i,83])&(know_2017.bq41_2!=0)].bq41_2.median())
    

In [38]:
null_li3 = know_2017[(know_2017.bq39_1==2) & (know_2017.bq41_2==0)].index
null_li3 

Int64Index([], dtype='int64')

In [39]:
## 자격증 항목 없음, 없다 처리 

know_2017['bq4_1b'].replace(0, '0', inplace = True)


for i in know_2017.bq4_1b:
    if '없' in i : 
        print(i)

        
know_2017['bq4_1b'].replace('없다', '0', inplace = True)

없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다


In [40]:

for i in know_2017.bq4_1b:
    if '없' in i : 
        print(i)

In [41]:

for i in know_2017.bq4_1c:
    if '없' in i : 
        print(i)

        
know_2017['bq4_1c'].replace('없다', '0', inplace = True)

없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다
없다


In [42]:
for i in know_2017.bq4_1c:
    if '없' in i : 
        print(i)


In [43]:
know_2017[(know_2017.bq38==1) & (know_2017.bq38_1!='0')]['bq38_1'].value_counts()

Series([], Name: bq38_1, dtype: int64)

In [44]:
know_2017[(know_2017.bq38.isin([2,3,4,5,6])) & (know_2017.bq38_1=='0')]['bq38_1'].value_counts()

Series([], Name: bq38_1, dtype: int64)

In [45]:
## 학과 없음, 없다 처리 


for i in know_2017.bq38_1:
    if '없' in i : 
        print(i)

        
know_2017['bq38_1'].replace('없다', '0', inplace = True)

없다
없다
없다


In [46]:
for i in know_2017.bq38_1:
    if '없' in i : 
        print(i)


In [47]:
know_2017.knowcode = know_2017.knowcode.astype(str)

In [48]:
know_2017.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9486 entries, 0 to 9485
Data columns (total 156 columns):
 #    Column    Dtype 
---   ------    ----- 
 0    idx       int32 
 1    aq1_1     int32 
 2    aq1_2     int32 
 3    aq2_1     int32 
 4    aq2_2     int32 
 5    aq3_1     int32 
 6    aq3_2     int32 
 7    aq4_1     int32 
 8    aq4_2     int32 
 9    aq5_1     int32 
 10   aq5_2     int32 
 11   aq6_1     int32 
 12   aq6_2     int32 
 13   aq7_1     int32 
 14   aq7_2     int32 
 15   aq8_1     int32 
 16   aq8_2     int32 
 17   aq9_1     int32 
 18   aq9_2     int32 
 19   aq10_1    int32 
 20   aq10_2    int32 
 21   aq11_1    int32 
 22   aq11_2    int32 
 23   aq12_1    int32 
 24   aq12_2    int32 
 25   aq13_1    int32 
 26   aq13_2    int32 
 27   aq14_1    int32 
 28   aq14_2    int32 
 29   aq15_1    int32 
 30   aq15_2    int32 
 31   aq16_1    int32 
 32   aq16_2    int32 
 33   aq17_1    int32 
 34   aq17_2    int32 
 35   aq18_1    int32 
 36   aq18_2    in

In [49]:
know_2017.to_csv("know_2017_enutf.csv", encoding='utf8', index=False)
know_2017.to_csv("know_2017_encp.csv", encoding='cp949', index=False)

In [50]:
df_cp = pd.read_csv("know_2017_encp.csv", encoding='cp949')
df_cp.info(verbose=1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9486 entries, 0 to 9485
Data columns (total 156 columns):
 #    Column    Dtype 
---   ------    ----- 
 0    idx       int64 
 1    aq1_1     int64 
 2    aq1_2     int64 
 3    aq2_1     int64 
 4    aq2_2     int64 
 5    aq3_1     int64 
 6    aq3_2     int64 
 7    aq4_1     int64 
 8    aq4_2     int64 
 9    aq5_1     int64 
 10   aq5_2     int64 
 11   aq6_1     int64 
 12   aq6_2     int64 
 13   aq7_1     int64 
 14   aq7_2     int64 
 15   aq8_1     int64 
 16   aq8_2     int64 
 17   aq9_1     int64 
 18   aq9_2     int64 
 19   aq10_1    int64 
 20   aq10_2    int64 
 21   aq11_1    int64 
 22   aq11_2    int64 
 23   aq12_1    int64 
 24   aq12_2    int64 
 25   aq13_1    int64 
 26   aq13_2    int64 
 27   aq14_1    int64 
 28   aq14_2    int64 
 29   aq15_1    int64 
 30   aq15_2    int64 
 31   aq16_1    int64 
 32   aq16_2    int64 
 33   aq17_1    int64 
 34   aq17_2    int64 
 35   aq18_1    int64 
 36   aq18_2    in

In [51]:
df_utf = pd.read_csv("know_2017_enutf.csv")
df_utf.info(verbose=1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9486 entries, 0 to 9485
Data columns (total 156 columns):
 #    Column    Dtype 
---   ------    ----- 
 0    idx       int64 
 1    aq1_1     int64 
 2    aq1_2     int64 
 3    aq2_1     int64 
 4    aq2_2     int64 
 5    aq3_1     int64 
 6    aq3_2     int64 
 7    aq4_1     int64 
 8    aq4_2     int64 
 9    aq5_1     int64 
 10   aq5_2     int64 
 11   aq6_1     int64 
 12   aq6_2     int64 
 13   aq7_1     int64 
 14   aq7_2     int64 
 15   aq8_1     int64 
 16   aq8_2     int64 
 17   aq9_1     int64 
 18   aq9_2     int64 
 19   aq10_1    int64 
 20   aq10_2    int64 
 21   aq11_1    int64 
 22   aq11_2    int64 
 23   aq12_1    int64 
 24   aq12_2    int64 
 25   aq13_1    int64 
 26   aq13_2    int64 
 27   aq14_1    int64 
 28   aq14_2    int64 
 29   aq15_1    int64 
 30   aq15_2    int64 
 31   aq16_1    int64 
 32   aq16_2    int64 
 33   aq17_1    int64 
 34   aq17_2    int64 
 35   aq18_1    int64 
 36   aq18_2    in